contracts/Committee.sol

```diff
@@ -0,0 +1,548 @@
+			(bool found, uint pos) = findTimedOutStandby();
+			if (found) {
+				_evict(pos); // evict timed-out
+			} else {
+				(bool lowestWeightFound, uint lowestWeightPos, uint256 lowestWeight) = findLowestWeightStandby();
+				_evict(lowestWeightPos); // evict lowest weight
+			}
+			_onTopologyModification();
+			newStandbySize = maxStandbys;
+		}
+	}
+
+	function _adjustPositionInTopology(uint pos) private returns (bool committeeChanged, bool standbysChanged) {
+		// TODO if a validator leaves committee it may be replaced by a timed-out, ready-for-committee standby
+		(uint newPos, uint prevCommitteeSize, uint newCommitteeSize, uint prevStandbySize, uint newStandbySize) = _repositionTopologyMember(pos);
+
+		bool inCommitteeBefore = pos < prevCommitteeSize;
+		bool inStandbyBefore = !inCommitteeBefore;
+
+		bool inCommitteeAfter = newPos < newCommitteeSize;
+		bool inStandbyAfter = !inCommitteeAfter;
+
+		committeeChanged = inCommitteeBefore || inCommitteeAfter;
+		if (committeeChanged) {
+			_notifyCommitteeChanged();
+		}
+
+		standbysChanged = inStandbyBefore || inStandbyAfter;
+		if (standbysChanged) {
+			_notifyStandbysChanged();
+		}
+	}
+
+	function _isQualifiedForTopologyByRank(address validator) private view returns (bool qualified, uint entryPos) {
+		// this assumes maxTopologySize > maxCommitteeSize, otherwise a non ready-for-committee validator may override one that is ready.
+		(qualified, entryPos) = _isQualifiedForCommitteeByRank(validator);
+		if (qualified) {
+			return (true, entryPos);
+		}
+		return _isQualifiedAsStandbyByRank(validator);
+	}
+
+	function _isQualifiedAsStandbyByRank(address validator) private view returns (bool qualified, uint entryPos) {
+		if (isReadyToSyncStale(validator)) {
+			return (false, 0);
+		}
+
+		(bool found, uint atPos) = findTimedOutStandby();
+		if (found) {
+			return (true, atPos);
+		}
+
+		uint standbyCount = topology.length - committeeSize;
+		if (standbyCount < maxStandbys) {
+			return (true, topology.length);
+		}
+
+		(bool foundLowest, uint lowestRankedStandbyPos, uint lowestRankedStandbyWeight) = findLowestWeightStandby();
+		if (foundLowest && lowestRankedStandbyWeight < getValidatorWeight(validator)) {
+			return (true, lowestRankedStandbyPos);
+		}
+
+		return (false, 0);
+	}
+
+	function isAboveCommitteeEntryThreshold(address validator) private view returns (bool) {
+		return _compareValidatorsByData(minimumAddress, minimumWeight, true, validator, getValidatorWeight(validator), true) < 1;
+	}
+
+	function _isQualifiedForCommitteeByRank(address validator) private view returns (bool qualified, uint entryPos) {
+		// this assumes maxTopologySize > maxCommitteeSize, otherwise a non ready-for-committee validator may override one that is ready.
+		if (!isReadyForCommittee(validator) || isReadyToSyncStale(validator)) {
+			return (false, 0);
+		}
+
+		if (committeeSize >= minCommitteeSize && !isAboveCommitteeEntryThreshold(validator)) {
+			return (false, 0);
```

Comment:
BUG - we may still allow this validator to enter in a min-committee-size edge case